src/app/cli/src/init/mina_run.ml

```diff
@@ -21,6 +21,123 @@ let snark_pool_list t =
+            | `Bootstrapping ->
+                [%log debug] "Node is bootstrapping, waiting 15s before retry" ;
+                let%bind () = after (Time.Span.of_sec 15.0) in
+                wait_for_best_tip ()
+            | `Active breadcrumb ->
+                let protocol_state =
+                  Transition_frontier.Breadcrumb.protocol_state breadcrumb
+                in
+                let consensus_state =
+                  Mina_state.Protocol_state.consensus_state protocol_state
+                in
+                let current_slot =
+                  Consensus.Data.Consensus_state.curr_global_slot
+                    consensus_state
+                in
+                let blockchain_length =
+                  Consensus.Data.Consensus_state.blockchain_length
+                    consensus_state
+                in
+                let state_hash =
+                  Transition_frontier.Breadcrumb.state_hash breadcrumb
+                in
+                if
+                  Mina_numbers.Global_slot_since_hard_fork.(
+                    current_slot >= slot_tx_end)
+                then (
+                  [%log debug]
+                    "Best tip has reached slot_tx_end. Current slot: %s, \
+                     target slot: %s, blockchain_length: %s, state_hash: %s"
+                    (Mina_numbers.Global_slot_since_hard_fork.to_string
+                       current_slot )
+                    (Mina_numbers.Global_slot_since_hard_fork.to_string
+                       slot_tx_end )
+                    (Mina_numbers.Length.to_string blockchain_length)
+                    (Mina_base.State_hash.to_base58_check state_hash) ;
+                  Deferred.unit )
+                else (
+                  [%log debug]
+                    "Best tip not yet at slot_tx_end. Current slot: %s, target \
+                     slot: %s, blockchain_length: %s, state_hash: %s. Waiting \
+                     15s before retry"
+                    (Mina_numbers.Global_slot_since_hard_fork.to_string
+                       current_slot )
+                    (Mina_numbers.Global_slot_since_hard_fork.to_string
+                       slot_tx_end )
+                    (Mina_numbers.Length.to_string blockchain_length)
+                    (Mina_base.State_hash.to_base58_check state_hash) ;
+                  let%bind () = after (Time.Span.of_sec 15.0) in
+                  wait_for_best_tip () )
+          in
+          let%bind () = wait_for_best_tip () in
+          [%log info] "Generating hardfork config in daemon config directory" ;
+          let%bind result =
+            Mina_lib.Hardfork_config.dump_reference_config
+              ~breadcrumb_spec:`Stop_slot ~directory_name:config.conf_dir mina
+          in
+          match result with
+          | Ok () ->
+              [%log info]
+                "Successfully generated hardfork config, shutting down daemon" ;
+              (* Shutdown like Stop_daemon *)
```

Comment:
Nit: looks like there's different shutdown implementation across the code, would be nice to have them. Probably leave for follow ups though.